doc/internals.rst

```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
+Your backend can also take as input a set of backend-specific keyword
+arguments. All these keyword arguments can be passed to
+:py:meth:`~xarray.open_dataset` grouped either via the ``backend_kwarg``
+parameter or explicitly using the syntax ``**kwargs``.
+
+Output
+^^^^^^
+The output of the backend `open_dataset` shall be an instance of
+Xarray py:class:`~xarray.Dataset` that implements the additional method ``close``,
+used by Xarray to ensure the related files are eventually closed.
+
+If you don't want to support the lazy loading, then the :py:class:`~xarray.Dataset`
+shall contain ``NumPy.arrays`` and your work is almost done.
+
+open_dataset_parameters
++++++++++++++++++++++++
+``open_dataset_parameters`` is the list of backend ``open_dataset`` parameters.
+It is not a mandatory parameter, and if the backend does not provide it
+explicitly, Xarray creates a list of them automatically by inspecting the
+backend signature.
+
+Xarray uses ``open_dataset_parameters`` only when it needs to select
+the **decoders** supported by the backend.
+
+If ``open_dataset_parameters`` is not defined, but ``**kwargs`` and ``*args`` have
+been passed to the signature, Xarray raises an error.
+On the other hand, if the backend provides the ``open_dataset_parameters``,
+then ``**kwargs`` and `*args`` can be used in the signature.
+
+However, this practice is discouraged unless there is a good reasons for using
+`**kwargs` or `*args`.
+
+guess_can_open
+++++++++++++++
+``guess_can_open`` is used to identify the proper engine to open your data
+file automatically in case the engine is not specified explicitly. If you are
+not interested in supporting this feature, you can skip this step since
+py:class:`~xarray.backend.common.BackendEntrypoint` already provides a default
+py:meth:`~xarray.backend.common BackendEntrypoint.guess_engine` that always returns ``False``.
+
+Backend ``guess_can_open`` takes as input the ``filename_or_obj`` parameter of
+Xarray :py:meth:`~xarray.open_dataset`, and returns a boolean.
+
+
+How to register a backend
++++++++++++++++++++++++++++
+
+Define in your setup.py (or setup.cfg) a new entrypoint with:
+
+- group: ``xarray.backend``
+- name: the name to be passed to :py:meth:`~xarray.open_dataset`  as ``engine``.
+- object reference: the reference of the class that you have implemented.
+
+See https://packaging.python.org/specifications/entry-points/#data-model
+for more information
+
+How to support Lazy Loading
++++++++++++++++++++++++++++
+If you want to make your backend effective with big datasets, then you should support
+the lazy loading.
+Basically, you shall replace the :py:class:`numpy.array` inside the variables with
+a custom class:
+
+.. ipython:: python
+    backend_array = YourBackendArray()
+    data = indexing.LazilyOuterIndexedArray(backend_array)
+    variable = Variable(..., data, ...)
+
+Where ``YourBackendArray``is a class that inherit from
```

Comment:
Can we add an example ``YourBackendArray`` in the beginning that shows the signature for ``open_dataset``, the optional attributes, and the use of optional ``kwargs``? It would make this document easier to understand